### DISCUSSION TITLE
Treat enum.Flag and enum.IntFlag differently from Enums

### DISCUSSION THREAD

**Discussion by snazzyfox** [pydantic/pydantic#3698] [Feature Request]
The `Flag` and `IntFlag` classes in the `enum` stdlib module are useful for compactly passing a large number of boolean flags as a bit array, which is used by a number of APIs.

For example, the python stdlib `re` module defines a regex flags used for matching:

```py
class RegexFlag(enum.IntFlag):
    ASCII = A = 1
    IGNORECASE = I = 2
    LOCALE = L = 4
    UNICODE = U = 8
    MULTILINE = M = 16
    DOTALL = S = 32
    VERBOSE = X = 64

flags = RegexFlag(10)   # usually written as re.UNICODE | re.IGNORECASE
RegexFlag.IGNORECASE in flags  # returns True
```

Since the `Flag` classes subclass `Enum`, pydantic treats them as standard discrete Enums, so it throws validation errors if the value has more than one bit set:
```py
class MatchRequest(BaseModel):
    regex: str
    flags: RegexFlag
    string: str

match1 = MatchRequest(regex: 'a', string='abc', flags=2)  # works because 2 is a explicitly defined value in the enum
match2 = MatchRequest(regex: 'a', string='abc', flags=3)  # works because Flag subclasses int
match3 = MatchRequest(regex: 'a', string='abc', flags='2')  # does now work because pydantic treats the Flag as an Enum value, instead of an int
```

I'd like to propose that for `Flag` values, Pydantic should validate them differently from enums, since they take any value that's a sum of zero or more of the enum values.

**Comment by koevgeny10** [comment]
I add a more specific proposal.
Versions:
* Python: 3.13
* Pydantic: 2.10.6

Minimal example:
```python
from enum import Flag

from pydantic import TypeAdapter


class A(Flag):
    a = 1
    aa = 2
    aaa = 4

a_adapter = TypeAdapter(A)
```
Pydantic creates json schema for Flag the same as for Enum.
```python
print(a_adapter.json_schema())
{'enum': [1, 2, 4], 'title': 'A', 'type': 'integer'}
```
But Flag has different meening and api. In this example, `A` can take any integer between 0 and 7 and pydantic agrees with that.
```python
print(a_adapter.validate_python(3))
A.a|aa
```

So i think that default json schema should be something like this:
```python
{'minimum': 0, 'maximum': 7, 'title': 'A', 'type': 'integer'}
```
The minimum and maximum can be obtained as follows: `A(0).value` and `~A(0).value`.

**Comment by koevgeny10** [comment]
As for original post, Pydantic already validates flags differently from enums and accepts any value that's a sum of zero or more of the flag's values. In your example
> ```python
> match3 = MatchRequest(regex: 'a', string='abc', flags='2')  # does now work because pydantic treats the Flag as an Enum value, instead of an int
> ```
this doesn't work not becouse of Pydantic but becouse of Python. `RegexFlag("2")` throws the ValueError

**Comment by snazzyfox** [comment]
@koevgeny10 Good to know, and you are right - pydantic does appear to accept sums of values (for `Flag`, but not `IntFlag`, interestingly). However, the error messaging for both of them is the one from Enum, which is probably what confused me. Here's a more complete example:

```py
class Foo(Flag):
  a = auto()
  b = auto()
  c = auto()

class FooModel(BaseModel):
  foo: Foo

FooModel(foo=1)  # OK - FooModel(foo=<Foo.a: 1>)
FooModel(foo=3) # OK -  FooModel(foo=<Foo.b|a: 3>)
FooModel(foo='1')  # Fail - Input should be 1, 2 or 4 [type=enum, input_value='1', input_type=str]
FooModel(foo='3') # Fail - Input should be 1, 2 or 4 [type=enum, input_value='3', input_type=str]
```

```py
class Bar(IntFlag):
  a = auto()
  b = auto()
  c = auto()

class BarModel(BaseModel):
  bar: Bar

BarModel(bar=1)  # OK - BarModel(bar=<Bar.a: 1>)
BarModel(bar=3) # OK -  BarModel(bar=<Bar.b|a: 3>)
BarModel(bar='1')  # OK - BarModel(bar=<Bar.a: 1>) - IntFlag gets treated as int and pre-cast by pydantic, sure
BarModel(bar='3') # Fail - Input should be 1, 2 or 4 [type=enum, input_value='3', input_type=str] - huh?
```

So it seems like there are two issues here:

### The validation errors on Flags is confusing

Flags probably should get their own validation errors instead of reusing the one from enums, because the message sounds like flags aren't supported at all, even though they are.

### Automatic casting works for IntFlag *only if it's a defined value*

This one feels more like a bug - IntFlag behaves like a Flag if an int is provided, but behaves like an IntEnum if it gets something else. Given that neither of them natively accept strings (`Bar('1')` is an error), this seems like something that needs to be aligned on pydantic's side.
